pkg/new_storage/cos.go

```diff
@@ -78,30 +78,48 @@ func (c *COS) DeleteFile(key string) error {
 }
 
 func (c *COS) Walk(cosPath string, recursive bool, process func(RemoteFile) error) error {
-	prefix := path.Join(c.Config.Path, cosPath)
+	// COS needs prefix ended with "/".
+	prefix := path.Join(c.Config.Path, cosPath) + "/"
+
 	delimiter := ""
+	if !recursive {
+		//
+		// When delimiter is "/", we only process all backups in the CommonPrefixes field of response.
+		// Then we get backupLists.
+		//
+		delimiter = "/"
+	} else {
+		//
+		// When delimiter is an empty string, we  process the items under specified path.
+		// Then we can Delete File object.
+		//
+		delimiter = ""
+	}
 	res, _, err := c.client.Bucket.Get(context.Background(), &cos.BucketGetOptions{
 		Delimiter: delimiter,
 		Prefix:    prefix,
 	})
 	if err != nil {
 		return err
 	}
+	// When recursive is false, only process all the backups in the CommonPrefixes part.
 	for _, dir := range res.CommonPrefixes {
 		if err := process(&cosFile{
 			name: strings.TrimPrefix(dir, prefix),
 		}); err != nil {
 			return err
 		}
 	}
-	for _, v := range res.Contents {
-		modifiedTime, _ := parseTime(v.LastModified)
-		if err := process(&cosFile{
-			name:         strings.TrimPrefix(v.Key, prefix),
-			lastModified: modifiedTime,
-			size:         int64(v.Size),
-		}); err != nil {
-			return err
+	if recursive {
```

Comment:
looks much better

Reply:
It is not ok. Passing bool value will turn function more complex.